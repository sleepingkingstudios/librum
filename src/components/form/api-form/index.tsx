import * as React from 'react';
import type { FormikValues } from 'formik';
import type { IconDefinition } from '@fortawesome/fontawesome-common-types';

import type {
  Refetch,
  Response,
} from '@api';
import type { Animations } from '@components/types';
import { Form } from '../form';
import type {
  FormStatus,
  OnSubmit,
} from '../types';
import {
  handleSubmit,
  mapStatus,
} from './utils';

export type ApiFormProps = {
  children: React.ReactNode,
  className?: string,
  initialValues: FormikValues,
  loadingAnimation?: Animations,
  loadingIcon?: IconDefinition,
  loadingMessage?: string,
  refetch: Refetch,
  response: Response,
};

export const ApiForm = ({
  refetch,
  response,
  ...rest
}: ApiFormProps): JSX.Element => {
  const { isLoading } = response;
  const status: FormStatus = mapStatus(response);
  const onSubmit: OnSubmit = handleSubmit(refetch);

  return (
    <Form
      initialStatus={status}
      isLoading={isLoading}
      onSubmit={onSubmit}
      {...rest}
    />
  );
};
