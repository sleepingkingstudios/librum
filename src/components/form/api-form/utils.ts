import type { FormikValues } from 'formik';

import type {
  Refetch,
  RefetchOptions,
  RequestBody,
  Response,
} from '@api';
import { invalidParametersError } from '@api/errors';
import type {
  FormStatus,
  InvalidParametersError,
  SubmitHandler,
  SubmitHandlerOptions,
} from '../types';

export const handleSubmit = (
  refetch: Refetch,
): SubmitHandler => async (
  values: FormikValues,
  options: SubmitHandlerOptions,
): Promise<Response> => {
  const { setStatus } = options;
  const body: RequestBody = values;
  const request: RefetchOptions = { body };

  const response: Response = await refetch(request);

  const status: FormStatus = mapStatus(response);

  setStatus(status);

  return new Promise(resolve => resolve(response));
};

export const mapStatus = (response: Response): FormStatus => {
  const { errorType, hasError, status } = response;

  if (status !== 'failure') { return { ok: true }; }

  if (!hasError || typeof response.error === 'string') { return { ok: false }; }

  if (errorType !== invalidParametersError) { return { ok: false }; }

  const error = response.error as InvalidParametersError;
  const { data } = error;
  const { errors } = data;

  return { ok: false, errors };
};
