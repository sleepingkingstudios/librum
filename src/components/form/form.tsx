import * as React from 'react';
import {
  Form as FormWrapper,
  Formik,
} from 'formik';
import type { IconDefinition } from '@fortawesome/fontawesome-common-types';

import { LoadingOverlay } from '@components/loading-overlay';
import type { Animations } from '@components/types';
import { joinClassNames } from '@utils/react-utils';
import type { FormProps } from './types';

const renderLoadingOverlay = ({
  isLoading,
  loadingAnimation,
  loadingIcon,
  loadingMessage,
}: {
  isLoading: boolean;
  loadingAnimation?: Animations;
  loadingIcon?: IconDefinition;
  loadingMessage?: string;
}): JSX.Element => {
  if (!isLoading) { return null; }

  return (
    <LoadingOverlay
      animate={loadingAnimation}
      icon={loadingIcon}
      message={loadingMessage}
    />
  );
};

export const Form = (
  {
    children,
    className = null,
    initialStatus,
    initialValues,
    isLoading = false,
    loadingAnimation,
    loadingIcon,
    loadingMessage = null,
    onSubmit,
  }: FormProps
): JSX.Element => {
  const joinedClassName = joinClassNames(
    className,
    'relative',
  );

  return (
    <Formik
      initialStatus={initialStatus}
      initialValues={initialValues}
      onSubmit={onSubmit}
    >
      <FormWrapper className={joinedClassName}>
        {
          renderLoadingOverlay({
            isLoading,
            loadingAnimation,
            loadingIcon,
            loadingMessage,
          })
        }

        { children }
      </FormWrapper>
    </Formik>
  );
};
