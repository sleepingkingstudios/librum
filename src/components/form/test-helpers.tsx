import * as React from 'react';
import {
  Form,
  Formik,
  FormikValues,
} from 'formik';

import type { OnSubmit } from './types';

interface IFormWrapperProps {
  initialValues?: FormikValues;
  onSubmit?: OnSubmit;
  submitButton?: true | false;
}

const SubmitButton = (
  { submitButton }: { submitButton: true | false }
): JSX.Element => {
  if (!submitButton) { return null; }

  return (
    <button type="submit">Submit</button>
  );
};

export const formWrapper = (
  { initialValues, onSubmit, submitButton }: IFormWrapperProps) => {
  const FormWrapper = (
    { children }: { children: React.ReactNode }
  ): JSX.Element => (
    <Formik initialValues={initialValues || {}} onSubmit={onSubmit}>
      <Form>
        { children }

        <SubmitButton submitButton={submitButton} />
      </Form>
    </Formik>
  );

  return FormWrapper;
};
