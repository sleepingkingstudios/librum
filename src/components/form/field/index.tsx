import * as React from 'react';
import {
  capitalize,
  last,
  map,
  snakeCase,
} from 'lodash';
import {
  useField,
  useFormikContext,
} from 'formik';
import type {
  FieldHookConfig,
  FieldInputProps,
} from 'formik';

import type { UpToTwelveColumns } from '@components/types';
import { joinClassNames } from '@utils/react-utils';
import { FormInput } from '../input';
import type { FormStatus } from '../types';
import { getServerErrors } from '../utils';

interface IFormFieldProps {
  className?: string;
  cols?: UpToTwelveColumns;
  id?: string;
  label?: string | false;
  name: string;
  type?: string;
}

interface IFormLabelProps {
  id: string;
  label: string | false;
}

interface IRenderComponentProps {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  field: FieldInputProps<any>;
  id?: string;
  type: string;
}

const columnSpanClassName = ({ cols }: { cols: number }): string | null => {
  if (cols === null) { return null; }

  return `col-span-${cols}`;
};

const convertNameToId = (name: string): string => (
  name.replace(/\]/g, '').split('[').map(snakeCase).join('_')
);

const convertNameToLabel = (name: string): string => {
  const segment: string = last(name.split(/\[/));
  const trimmed: string = snakeCase(segment.replace(/\]/, ''));

  return map(trimmed.split('_'), capitalize).join(' ');
};

const renderComponent = ({
  field,
  id = null,
  type,
}: IRenderComponentProps): JSX.Element => (
  <FormInput field={field} id={id} type={type} />
);

const renderErrors = ({ errors }: { errors: string[] }): JSX.Element => {
  if (errors.length === 0) { return; }

  return (<FormFieldErrors errors={errors} />);
};

const renderLabel = ({ id, label }: IFormLabelProps): JSX.Element => {
  if (label === false) {
    return null;
  }

  return (
    <label className="form-label" htmlFor={id}>{ label }</label>
  );
};

const FormFieldErrors = ({ errors }: { errors: string[] }): JSX.Element => (
  <span className="form-field-errors">{ errors.join(', ') }</span>
);

export const FormField = ({
  className = null,
  cols = null,
  id = null,
  label = null,
  name,
  type = 'text',
}: IFormFieldProps): JSX.Element => {
  const configuredId = id || convertNameToId(name);
  const configuredLabel =
    label === false ? label : (label || convertNameToLabel(name));
  const joinedClassName = joinClassNames(
    'form-field',
    columnSpanClassName({ cols }),
    className,
  );
  const fieldConfig: FieldHookConfig<string> = {
    name,
  };
  const [field] = useField(fieldConfig);
  const context = useFormikContext();
  const status = context.status as FormStatus;
  const errors = getServerErrors({ name, status });

  return (
    <div className={joinedClassName}>
      { renderLabel({ id: configuredId, label: configuredLabel }) }

      { renderComponent({ id: configuredId, field, type }) }

      { renderErrors({ errors }) }
    </div>
  );
};
