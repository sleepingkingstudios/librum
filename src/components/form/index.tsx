import * as React from 'react';
import {
  Form as FormWrapper,
  Formik,
  FormikValues,
} from 'formik';
import type { IconDefinition } from '@fortawesome/fontawesome-common-types';

import type {
  Response,
  UseMutationTrigger,
} from '@api';
import { LoadingOverlay } from '@components/loading-overlay';
import type { Animations } from '@components/types';
import { joinClassNames } from '@utils/react-utils';
import { handleSubmit } from './utils';

export { FormButton } from './button';
export { FormField } from './field';
export { FormInput } from './input';
export { FormRow } from './row';
export { FormSubmitButton } from './submit-button';

interface FormProps {
  children: React.ReactNode;
  className?: string;
  initialValues?: FormikValues;
  loadingAnimation?: Animations;
  loadingIcon?: IconDefinition;
  loadingMessage?: string;
  request: UseMutationTrigger;
  response: Response;
}

interface RenderLoadingOverlay {
  isLoading: boolean;
  loadingAnimation?: Animations;
  loadingIcon?: IconDefinition;
  loadingMessage?: string;
}

const renderLoadingOverlay = ({
  isLoading,
  loadingAnimation,
  loadingIcon,
  loadingMessage,
}: RenderLoadingOverlay): JSX.Element => {
  if (!isLoading) { return null; }

  return (
    <LoadingOverlay
      animate={loadingAnimation}
      icon={loadingIcon}
      message={loadingMessage}
    />
  );
};

export const Form = (
  {
    children,
    className = null,
    initialValues,
    loadingAnimation,
    loadingIcon,
    loadingMessage = null,
    request,
    response,
  }: FormProps
): JSX.Element => {
  const { isLoading } = response;
  const joinedClassName = joinClassNames(
    className,
    'relative',
  );

  return (
    <Formik initialValues={initialValues} onSubmit={handleSubmit(request)}>
      <FormWrapper className={joinedClassName}>
        {
          renderLoadingOverlay({
            isLoading,
            loadingAnimation,
            loadingIcon,
            loadingMessage,
          })
        }

        { children }
      </FormWrapper>
    </Formik>
  );
};
