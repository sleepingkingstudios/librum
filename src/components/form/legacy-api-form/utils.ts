import type { FormikValues } from 'formik';

import type {
  ApiFailure,
  UseMutationTrigger,
} from '@api';
import { invalidParametersError } from '@api/errors';
import { camelizeErrorType } from '@api/hooks/utils';
import { convertToCamelCase } from '@utils/data';
import type {
  FormErrors,
  InvalidParametersError,
  SubmitHandler,
  SubmitHandlerOptions,
  SubmitResponse,
} from '../types';

const extractErrors = (response: ApiFailure): FormErrors => {
  if (camelizeErrorType(response.error.type) !== invalidParametersError) {
    return null;
  }

  const error = response.error as InvalidParametersError;
  const { data } = error;
  const { errors } = data;

  return convertToCamelCase(errors) as FormErrors;
};

export const handleSubmit = (
  request: UseMutationTrigger,
): SubmitHandler => async (
  values: FormikValues,
  options: SubmitHandlerOptions,
): Promise<void> => {
  const { setStatus } = options;
  const response = await request(values) as SubmitResponse;

  if (!response || !('error' in response)) {
    setStatus({ ok: true });

    return;
  }

  const { error } = response;

  if (!('data' in error) || !('status' in error) || !(typeof error.status === 'number')) {
    setStatus({ ok: false });

    return;
  }

  const data = error.data as ApiFailure;
  const errors = extractErrors(data);

  if (!errors) {
    setStatus({ ok: false });

    return;
  }

  setStatus({ ok: false, errors });
};
