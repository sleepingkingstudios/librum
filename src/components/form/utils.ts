import type { FormikValues } from 'formik';

import type {
  ApiFailure,
  UseMutationTrigger,
} from '@api';
import { invalidParametersError } from '@api/errors';
import { camelizeErrorType } from '@api/hooks/utils';
import { convertToCamelCase } from '@utils/data';
import type {
  FormError,
  FormErrors,
  FormStatus,
  InvalidParametersError,
  SubmitHandler,
  SubmitHandlerOptions,
  SubmitResponse,
} from './types';

const extractErrors = (response: ApiFailure): FormErrors => {
  if (camelizeErrorType(response.error.type) !== invalidParametersError) {
    return null;
  }

  const error = response.error as InvalidParametersError;
  const { data } = error;
  const { errors } = data;

  return convertToCamelCase(errors) as FormErrors;
};

export const getServerErrors = ({
  name,
  status,
}: {
  name: string,
  status: FormStatus,
}): string[] => {
  if (!status) { return []; }

  if (status.ok) { return []; }

  if (!('errors' in status)) { return []; }

  const { errors } = status;
  const fieldErrors = (errors[name] || []) as FormError[];
  const messages = fieldErrors.map(error => error.message);

  return messages;
};

export const handleSubmit = (
  request: UseMutationTrigger,
): SubmitHandler => async (
  values: FormikValues,
  options: SubmitHandlerOptions,
): Promise<void> => {
  const { setStatus } = options;
  const response = await request(values) as SubmitResponse;

  if (!response || !('error' in response)) {
    setStatus({ ok: true });

    return;
  }

  const { error } = response;

  if (!('data' in error) || !('status' in error) || !(typeof error.status === 'number')) {
    setStatus({ ok: false });

    return;
  }

  const data = error.data as ApiFailure;
  const errors = extractErrors(data);

  if (!errors) {
    setStatus({ ok: false });

    return;
  }

  setStatus({ ok: false, errors });
};
