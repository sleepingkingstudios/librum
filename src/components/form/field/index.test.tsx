import * as React from 'react';

import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';

import { FormField } from './index';
import { formWrapper } from '@components/form/test-helpers';
import type {
  OnSubmit,
  SubmitHandler,
} from '../types';
import { getServerErrors } from '../utils';

jest.mock('../utils');

const mockGetServerErrors = getServerErrors as jest.MockedFunction<typeof getServerErrors>;

mockGetServerErrors.mockImplementation(() => []);

describe('<FormField />', () => {
  const options = { setStatus: jest.fn() };
  const submitHandler: jest.MockedFunction<SubmitHandler> = jest.fn();
  const onSubmit: OnSubmit =
    (values) => submitHandler(values, options);

  beforeEach(() => { submitHandler.mockClear(); });

  it('should pass the value to the form', async () => {
    const initialValues = { email: '' };

    const FormWrapper = formWrapper({
      initialValues,
      onSubmit,
      submitButton: true,
    });
    const { getByRole } = render(
      <FormField name="email" />,
      { wrapper: FormWrapper },
    );

    await userEvent.click(getByRole('button', { name: 'Submit'}));

    expect(submitHandler).toHaveBeenCalledWith(initialValues, options);
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <FormField name="email" />,
      { wrapper: formWrapper({ onSubmit: jest.fn() }) },
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('when the form has data for the input', () => {
    it('should show the value in the input', () => {
      const initialValues = { email: 'alan.bradley@example.com' };

      const FormWrapper = formWrapper({
        initialValues,
        onSubmit: jest.fn(),
        submitButton: true,
      });
      const { getByRole } = render(
        <FormField name="email" />,
        { wrapper: FormWrapper },
      );
      const input = getByRole('textbox');

      expect(input).toBeVisible();
      expect(input).toHaveValue(initialValues['email']);
    });

    it('should pass the value to the form', async () => {
      const initialValues = { email: 'alan.bradley@example.com' };

      const FormWrapper = formWrapper({
        initialValues,
        onSubmit,
        submitButton: true,
      });
      const { getByRole } = render(
        <FormField name="email" />,
        { wrapper: FormWrapper },
      );

      await userEvent.click(getByRole('button', { name: 'Submit'}));

      expect(submitHandler).toHaveBeenCalledWith(initialValues, options);
    });

    it('should match the snapshot', () => {
      const initialValues = { email: 'alan.bradley@example.com' };
      const FormWrapper = formWrapper({
        initialValues,
        onSubmit: jest.fn(),
      });
      const { asFragment } = render(
        <FormField name="email" />,
        { wrapper: FormWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when the form has server errors', () => {
    const errors = ["can't be blank", 'invalid format'];

    beforeEach(() => {
      mockGetServerErrors.mockImplementation(() => errors);
    });

    afterEach(() => {
      mockGetServerErrors.mockImplementation(() => []);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormField name="email" />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when the user enters text', () => {
    it('should show the value in the input', async () => {
      const initialValues = { email: '' };

      const FormWrapper = formWrapper({
        initialValues,
        onSubmit: jest.fn(),
        submitButton: true,
      });
      const { getByRole } = render(
        <FormField name="email" />,
        { wrapper: FormWrapper },
      );
      const input = getByRole('textbox');

      expect(input).toBeVisible();

      await userEvent.type(input, 'alan.bradley@example.com');

      expect(input).toHaveValue('alan.bradley@example.com');
    });

    it('should pass the value to the form', async () => {
      const initialValues = { email: '' };

      const FormWrapper = formWrapper({
        initialValues,
        onSubmit,
        submitButton: true,
      });
      const { getByRole } = render(
        <FormField name="email" />,
        { wrapper: FormWrapper },
      );

      await userEvent.type(getByRole('textbox'), 'alan.bradley@example.com');

      await userEvent.click(getByRole('button', { name: 'Submit'}));

      expect(submitHandler).toHaveBeenCalledWith({
        email: 'alan.bradley@example.com'
      },
      options);
    });
  });

  describe('with className: value', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormField name="email" className="custom-form-field" />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with cols: value', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormField name="email" cols={3} />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with id: value', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormField id="user-email" name="email" />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with label: false', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormField name="email" label={false} />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with label: value', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormField name="email" label="Your Email" />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with name: qualified value in camelCase', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormField name="user[contact][emailAddress]" />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with name: qualified value in snake_case', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormField name="user[contact][email_address]" />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with type: password', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormField name="email" type="password" />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with type: text', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormField name="email" type="text" />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
