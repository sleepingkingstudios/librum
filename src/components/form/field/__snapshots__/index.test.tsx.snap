// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormField /> should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field"
    >
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        class="form-input"
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`<FormField /> when the form has data for the input should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field"
    >
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        class="form-input"
        id="email"
        name="email"
        type="text"
        value="alan.bradley@example.com"
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`<FormField /> when the form has server errors should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field"
    >
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        class="form-input"
        id="email"
        name="email"
        type="text"
        value=""
      />
      <span
        class="form-field-errors"
      >
        can't be blank, invalid format
      </span>
    </div>
  </form>
</DocumentFragment>
`;

exports[`<FormField /> with className: value should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field custom-form-field"
    >
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        class="form-input"
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`<FormField /> with cols: value should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field col-span-3"
    >
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        class="form-input"
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`<FormField /> with id: value should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field"
    >
      <label
        class="form-label"
        for="user-email"
      >
        Email
      </label>
      <input
        class="form-input"
        id="user-email"
        name="email"
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`<FormField /> with label: false should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field"
    >
      <input
        class="form-input"
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`<FormField /> with label: value should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field"
    >
      <label
        class="form-label"
        for="email"
      >
        Your Email
      </label>
      <input
        class="form-input"
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`<FormField /> with name: qualified value in camelCase should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field"
    >
      <label
        class="form-label"
        for="user_contact_email_address"
      >
        Email Address
      </label>
      <input
        class="form-input"
        id="user_contact_email_address"
        name="user[contact][emailAddress]"
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`<FormField /> with name: qualified value in snake_case should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field"
    >
      <label
        class="form-label"
        for="user_contact_email_address"
      >
        Email Address
      </label>
      <input
        class="form-input"
        id="user_contact_email_address"
        name="user[contact][email_address]"
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`<FormField /> with type: password should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field"
    >
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        class="form-input"
        id="email"
        name="email"
        type="password"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`<FormField /> with type: text should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form-field"
    >
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        class="form-input"
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;
